istioctl/cmd/istioctl/istioctl_test.go

```diff
@@ -0,0 +1,577 @@
+			`Deleted config: routerule a
+Deleted config: routerule b
+`,
+			regexp.MustCompile("^$"),
+			false,
+		},
+		{ // case 3 - delete by filename of istio config which doesn't exist
+			testRouteRules,
+			strings.Split("delete -f convert/testdata/v1alpha1/route-rule-80-20.yaml", " "),
+			"",
+			regexp.MustCompile("^Error: 1 error occurred:\n\n\\* cannot delete route-rule/default/route-rule-80-20: item not found\n$"),
+			true,
+		},
+	}
+
+	for i, c := range cases {
+		// Override the client factory used by main.go
+		clientFactory = mockClientFactoryGenerator(c.configs)
+
+		// run deleteCmd capturing output
+		var out bytes.Buffer
+		rootCmd.SetOutput(&out)
+		rootCmd.SetArgs(c.args)
+		// fErr := rootCmd.Execute()
+
+		file = "" // Clear, because we re-use
+
+		stdOutput, fErr := captureStdout(
+			func() error {
+				rootCmd.SetArgs(c.args)
+				err := rootCmd.Execute()
+				if err != nil {
+					return multierror.Prefix(err, fmt.Sprintf("Could not run command %v", c.args))
+				}
+				return nil
+			})
+		errOutput := out.String()
+
+		if c.wantOutput != stdOutput {
+			t.Errorf("Stdout didn't match for delete case %d \"istioctl delete %v\": Expected %q, got %q.  Stderr was %q",
+				i, strings.Join(c.args, " "),
+				c.wantOutput, stdOutput, errOutput)
+		}
+
+		if !c.wantError.MatchString(errOutput) {
+			t.Errorf("Stderr didn't match for delete case %d %v: Expected %v, got %q", i, c.args, c.wantError, errOutput)
+		}
+
+		if c.wantException {
+			if fErr == nil {
+				t.Errorf("Wanted an exception for case %d %v, didn't get one, output was %q", i, c.args, stdOutput)
+			}
+		} else {
+			if fErr != nil {
+				t.Errorf("Unwanted exception for case %d %v: %v", i, c.args, fErr)
+			}
+		}
+	}
+}
+
+// mockClientFactoryGenerator creates a factory for model.ConfigStore preloaded with data
+func mockClientFactoryGenerator(configs []model.Config) func() (model.ConfigStore, error) {
+	outFactory := func() (model.ConfigStore, error) {
+		// Initialize the real client to get the supported config types
+		realClient, err := newClient()
+		if err != nil {
+			return nil, err
+		}
+
+		// Initialize memory based model.ConfigStore with configs
+		outConfig := memory.Make(realClient.ConfigDescriptor())
+		for _, config := range configs {
+			if _, err := outConfig.Create(config); err != nil {
+				return nil, err
+			}
+		}
+
+		// Wrap the memory ConfigStore so List() is sorted
+		return sortedConfigStore{store: outConfig}, nil
+	}
+
+	return outFactory
+}
+
+// captureStdout invokes f capturing the output sent to stderr and stdout
+func captureStdout(f func() error) (string, error) {
+	origStdout := os.Stdout
+	rOut, wOut, _ := os.Pipe()
+	os.Stdout = wOut
+
+	errF := f()
+
+	outChannel := make(chan string)
+	go func() {
+		var buf bytes.Buffer
+		io.Copy(&buf, rOut)
+		outChannel <- buf.String()
+	}()
+
+	wOut.Close()
+	os.Stdout = origStdout
+	capturedOutput := <-outChannel
+
+	return capturedOutput, errF
+}
```

Comment:
A quick google search for this pattern indicates you may want to start the goroutine _before_ invoking the test function (see [here](https://stackoverflow.com/questions/10473800/in-go-how-do-i-capture-stdout-of-a-function-into-a-string?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa#comment46866149_10476304)). The golang testing example code might be a good reference (see  https://golang.org/src/testing/example.go#L60).

Also, I'm not sure how capturing stdio will work if tests are later run in parallel. It might be good to convert istioctl functions over to use explicit `printf` and `fatalf` interface functions like Mixer and Galley code. We could then redirect that to cobra.Print or provide our own fake to explicit capture all stdout and stderr.